src/ui/utils.ts

```diff
@@ -1,7 +1,161 @@
-import {isMobile} from '../utils/platform';
+import {isMobile, isFirefox} from '../utils/platform';
+import UserStorage from '../background/user-storage';
+import {getSRGBLightness, parse, RGBA} from '../utils/color';
 
 declare const __CHROMIUM_MV3__: boolean;
 
+declare const browser: {
+    theme: {
+        getCurrent: (() => Promise<void>);
+    };
+};
+
+export async function setTheme() {
+    // Check if theme setting should be used
+    if (isFirefox && typeof browser !== 'undefined' && browser.theme && browser.theme.getCurrent) {
+        await UserStorage.loadSettings();
+        if (UserStorage.settings.useFirefoxTheme) {
+            applyTheme(await browser.theme.getCurrent());
+        }
+    }
+}
+
+function blendColors(background: RGBA, color: RGBA, alpha: number): RGBA {
+    // alpha between 0 and 1
+
+    return {
+        r: Math.floor(background.r + alpha * (color.r - background.r)),
+        g: Math.floor(background.g + alpha * (color.g - background.g)),
+        b: Math.floor(background.b + alpha * (color.b - background.b)),
+        a: 1};
+}
+
+function applyTheme(theme: any) {
+    if (theme.colors) {
+
+        // Get text and background colors
+        const text = theme.colors.popup_text;
+        const background = theme.colors.popup;
+
+        // Parse them into RGBA
+        const backgroundRGBA = parse(background);
+        const textRGBA = parse(text);
+
+        // If a value is missing, exit
+        if (backgroundRGBA == null || textRGBA == null) {
+            return;
+        }
+
+        // Check whether theme is light or dark
+        const light = Math.round(getSRGBLightness(backgroundRGBA.r, backgroundRGBA.g, backgroundRGBA.b));
+
+        // Create different highlights
+        const blendWeak = blendColors(backgroundRGBA, textRGBA, .17);
+        const highlightWeak = `rgb(${blendWeak.r}, ${blendWeak.g}, ${blendWeak.b})`;
+        const blendStrong = blendColors(backgroundRGBA, textRGBA, .25);
+        const highlightStrong = `rgb(${blendStrong.r}, ${blendStrong.g}, ${blendStrong.b})`;
+
+        // Create and apply css
+        const style = document.createElement('style');
+        style.textContent = `
+        * {
+            color: ${text} !important;
+            border-color: ${highlightWeak} !important;
+            scrollbar-color: ${text} ${highlightWeak} !important;
+        }
+
+        /* Icon inversions */
+        .settings-button-icon, .m-help-button__text::before, .nav-button::after,
```

Comment:
Sorry, but I cannot merge this implementation. I have some ideas about changing the UI theme dynamically, and when I get to refactoring the use of color variables it will be easier for you to implement your option.